src/controller/CHIPDeviceController.cpp

```diff
@@ -1588,43 +1588,59 @@ void DeviceCommissioner::SetupCluster(ClusterBase & base, DeviceProxy * proxy, E
     base.SetCommandTimeout(timeout);
 }
 
-void BasicVendorCallback(void * context, VendorId vendorId)
+// AttributeCache::Callback impl
+void DeviceCommissioner::OnDone()
 {
-    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
-    CommissioningDelegate::CommissioningReport report;
-    report.Set<BasicVendor>(vendorId);
-    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
-}
+    CHIP_ERROR err;
+    CHIP_ERROR return_err = CHIP_NO_ERROR;
+    ReadCommissioningInfo info;
 
-void BasicProductCallback(void * context, uint16_t productId)
-{
-    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
-    CommissioningDelegate::CommissioningReport report;
-    report.Set<BasicProduct>(productId);
-    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
-}
+    // Using ForEachAttribute because this attribute can be queried on any endpoint.
+    err = mAttributeCache->ForEachAttribute(
+        app::Clusters::GeneralCommissioning::Id, [this, &info](const app::ConcreteAttributePath & path) {
```

Comment:
For what it's worth, the device library spec does define that General Commissioning has to be on endpoint 0 and only endpoint 0.

Reply:
Root node device type needs to be on EP0, but it doesn't seem to preclude having a general commissioning cluster on other endpoints - similar to the argument that was made for having multiple networking clusters. It's less likely, but I don't think it's banned.

Reply:
It absolutely does preclude it.  See device library spec.  The "Genercal Commissioning" cluster has the "I" quality, which means "singleton".  "Network Commissioning", on the other hand, does _not_ have that quality.